Confusing test failures on free-threaded (1) seem to report that the arrays differ on a relative scale of inf. But the printed values for both arrays agree. Is there a good way to track down these errors (perhaps it is not thread safe in some way)? Perhaps it is checking for array agreement before all threads are completed, and the report of the failure is done after the threads are all complete?

```
_______________ ERROR at call of test_sum[shape22-axis22-out22] ________________
scipy/sparse/tests/test_coo.py:962: in test_sum
    assert_equal(res, exp)
E   AssertionError: 
E   Arrays are not equal
E   
E   Mismatched elements: 5 / 5 (100%)
E   Max absolute difference among violations: 6.98210571e+19
E   Max relative difference among violations: inf
E    ACTUAL: array([ 5.369791e+19,  1.334737e+19,  1.093169e+19,  6.982106e+19,
E          -6.520094e+18])
E    DESIRED: array([ 5.369791e+19,  1.334737e+19,  1.093169e+19,  6.982106e+19,
E          -6.520094e+18])
        a          = <COOrdinate sparse array of dtype 'int64'
	with 504 stored elements and shape (3, 7, 8, 5)>
        axis       = (0, 1, 2)
        exp        = array([ 5.36979055e+19,  1.33473720e+19,  1.09316873e+19,  6.98210571e+19,
       -6.52009378e+18])
        out        = array([ 5.36979055e+19,  1.33473720e+19,  1.09316873e+19,  6.98210571e+19,
       -6.52009378e+18])
        res        = array([ 5.36979055e+19,  1.33473720e+19,  1.09316873e+19,  6.98210571e+19,
       -6.52009378e+18])
        rng        = Generator(PCG64) at 0x339D06EE380
        shape      = (3, 7, 8, 5)
```

--- a/scipy/sparse/_base.py
+++ b/scipy/sparse/_base.py
@@ -5,7 +5,7 @@ import numpy as np
 
 from ._sputils import (asmatrix, check_reshape_kwargs, check_shape,
                        get_sum_dtype, isdense, isscalarlike, _todata,
-                       matrix, validateaxis, getdtype, isintlike)
+                       matrix, validateaxis, getdtype)
 from scipy._lib._sparse import SparseABC, issparse
 
 from ._matrix import spmatrix

--- a/scipy/sparse/_base.py
+++ b/scipy/sparse/_base.py
@@ -4,7 +4,7 @@ import math
 import numpy as np
 
 from ._sputils import (asmatrix, check_reshape_kwargs, check_shape,
-                       get_sum_dtype, isdense, isscalarlike,
+                       get_sum_dtype, isdense, isscalarlike, _todata,
                        matrix, validateaxis, getdtype, isintlike)
 from scipy._lib._sparse import SparseABC, issparse
 
@@ -1148,25 +1148,39 @@ class _spbase(SparseABC):
         # Mimic numpy's casting.
         res_dtype = get_sum_dtype(self.dtype)
 
-        # Note: all valid 1D axis values are canonically `None` so use this code.
+        # Note: all valid 1D axis values are canonically `None`.
         if axis is None:
-            ones = self._ascontainer(np.ones((self.shape[-1], 1), dtype=res_dtype))
-            return (self @ ones).sum(dtype=dtype, out=out)
+            if self.nnz == 0:
+                return np.sum(self._ascontainer([0]), dtype=dtype or res_dtype, out=out)
+            return np.sum(self._ascontainer(_todata(self)), dtype=dtype, out=out)
+        elif isspmatrix(self):
+            # Ensure spmatrix sums stay 2D
+            new_shape = (1, self.shape[1]) if axis == (0,) else (self.shape[0], 1)
+        else:
+            new_shape = tuple(self.shape[i] for i in range(self.ndim) if i not in axis)
 
-        # We use multiplication by a matrix of ones to achieve this.
+        if out is None:
+            # create out array with desired dtype
+            out = np.zeros(new_shape, dtype=dtype or res_dtype)
+        else:
+            if out.shape != new_shape:
+                raise ValueError("out dimensions do not match shape")
+
+        if self.ndim > 2:
+            return self._sum_nd(axis, res_dtype, out)
+
+        # We use multiplication by a matrix of ones to sum.
         # For some sparse array formats more efficient methods are
         # possible -- these should override this function.
-        if axis == 0:
-            # sum over columns
+        if axis == (0,):
             ones = self._ascontainer(np.ones((1, self.shape[0]), dtype=res_dtype))
-            ret = ones @ self
-        else:  # axis == 1:
-            # sum over rows
+            # sets dtype while loading into out
+            out[...] = (ones @ self).reshape(new_shape)
+        else:  # axis == (1,)
             ones = self._ascontainer(np.ones((self.shape[1], 1), dtype=res_dtype))
-            ret = self @ ones
-
-        # This doesn't sum anything. It handles dtype and out.
-        return ret.sum(axis=axis, dtype=dtype, out=out)
+            # sets dtype while loading into out
+            out[...] = (self @ ones).reshape(new_shape)
+        return out
 
     def mean(self, axis=None, dtype=None, out=None):
         """
@@ -1207,32 +1221,21 @@ class _spbase(SparseABC):
         """
         axis = validateaxis(axis, ndim=self.ndim)
 
-        res_dtype = self.dtype.type
         integral = (np.issubdtype(self.dtype, np.integer) or
                     np.issubdtype(self.dtype, np.bool_))
 
-        # output dtype
-        if dtype is None:
-            if integral:
-                res_dtype = np.float64
-        else:
-            res_dtype = np.dtype(dtype).type
-
         # intermediate dtype for summation
-        inter_dtype = np.float64 if integral else res_dtype
+        inter_dtype = np.float64 if integral else self.dtype
         inter_self = self.astype(inter_dtype)
 
         if axis is None:
             denom = math.prod(self.shape)
-        elif isintlike(axis):
-            denom = self.shape[axis]
         else:
             denom = math.prod(self.shape[ax] for ax in axis)
-        res = (inter_self * (1.0 / denom)).sum(axis=axis, dtype=inter_dtype)
-        if out is None:
-            return res.sum(axis=(), dtype=dtype)
-        # out is handled differently by matrix and ndarray so use as_container
-        return self._ascontainer(res).sum(axis=(), dtype=dtype, out=out)
+        res = (inter_self * (1.0 / denom)).sum(axis=axis, dtype=inter_dtype, out=out)
+        if dtype is not None and out is None:
+            return res.astype(dtype, copy=False)
+        return res
 
 
     def diagonal(self, k=0):

--- a/scipy/sparse/_coo.py
+++ b/scipy/sparse/_coo.py
@@ -919,7 +919,7 @@ class _coo_base(_data_matrix, _minmax_mixin):
         else:
             o_new_shape = other.shape[:-2] + other.shape[-1:]
             reorder_dims = (o_ndim - 2, *range(o_ndim - 2), o_ndim - 1)
-            o_reorg = np.permute_dims(other, reorder_dims)
+            o_reorg = np.transpose(other, reorder_dims)
             other_2d = o_reorg.reshape((other.shape[-2], math.prod(o_new_shape)))
 
         prod = self_2d @ other_2d  # routes via 2-D CSR

--- a/scipy/sparse/_coo.py
+++ b/scipy/sparse/_coo.py
@@ -298,7 +298,7 @@ class _coo_base(_data_matrix, _minmax_mixin):
             M, N = self.shape
             coo_todense(M, N, self.nnz, self.row, self.col, self.data,
                         B.ravel('A'), fortran)
-        else:
+        else:  # dim>2
             if fortran:
                 strides = np.append(1, np.cumprod(self.shape[:-1]))
             else:
@@ -603,7 +603,6 @@ class _coo_base(_data_matrix, _minmax_mixin):
         A = self.__class__((new_data, new_coords), shape=self.shape)
         return A
 
-
     def _sub_sparse(self, other):
         if self.ndim < 3:
             return self.tocsr()._sub_sparse(other)
@@ -615,7 +614,6 @@ class _coo_base(_data_matrix, _minmax_mixin):
         new_coords = tuple(np.concatenate((self.coords, other.coords), axis=1))
         A = coo_array((new_data, new_coords), shape=self.shape)
         return A
-
 
     def _matmul_vector(self, other):
         if self.ndim > 2:
@@ -650,7 +648,6 @@ class _coo_base(_data_matrix, _minmax_mixin):
             return result[0]
         return result
 
-
     def _rmatmul_dispatch(self, other):
         if isscalarlike(other):
             return self._mul_scalar(other)
@@ -675,7 +672,6 @@ class _coo_base(_data_matrix, _minmax_mixin):
             else:
                 perm = tuple(range(ret.ndim)[:-2]) + tuple(range(ret.ndim)[-2:][::-1])
             return ret.transpose(perm)
-
 
     def _matmul_dispatch(self, other):
         if isscalarlike(other):
@@ -726,7 +722,6 @@ class _coo_base(_data_matrix, _minmax_mixin):
                     f"{err_prefix} (n,..,k={N}),(k={other.shape[-2]},..,m)->(n,..,m)"
                 )
 
-
         if isscalarlike(other):
             # scalar value
             return self._mul_scalar(other)
@@ -771,7 +766,6 @@ class _coo_base(_data_matrix, _minmax_mixin):
                 result = result.reshape(result.shape[:-1])
             return result
 
-
     def _matmul_multivector(self, other):
         result_dtype = upcast_char(self.dtype.char, other.dtype.char)
         if self.ndim >= 3 or other.ndim >= 3:
@@ -806,7 +800,6 @@ class _coo_base(_data_matrix, _minmax_mixin):
         coo_matmat_dense(self.nnz, other.shape[-1], row, col,
                          self.data, other.ravel('C'), result)
         return result.view(type=type(other))
-
 
     def dot(self, other):
         """Return the dot product of two arrays.
@@ -861,6 +854,7 @@ class _coo_base(_data_matrix, _minmax_mixin):
         >>> A.dot(B).shape
         (3, 4, 5, 5, 4, 3)
         """
+        # handle non-array input:  lists, ints, etc
         if not (issparse(other) or isdense(other) or isscalarlike(other)):
             # If it's a list or whatever, treat it like an array
             o_array = np.asanyarray(other)
@@ -872,118 +866,68 @@ class _coo_base(_data_matrix, _minmax_mixin):
             except AttributeError:
                 other = o_array
 
-        if self.ndim < 3 and (np.isscalar(other) or other.ndim<3):
-            return _spbase.dot(self, other)
         # Handle scalar multiplication
-        if np.isscalar(other):
+        if isscalarlike(other):
             return self * other
+
+        # other.shape[-2:][0] gets last index of 1d, next to last index of >1d
+        if self.shape[-1] != other.shape[-2:][0]:
+            raise ValueError(f"shapes {self.shape} and {other.shape}"
+                             " are not aligned for n-D dot")
+
+        if self.ndim < 3 and other.ndim < 3:
+            return self @ other
         if isdense(other):
             return self._dense_dot(other)
-        elif other.format != "coo":
-            raise TypeError("input must be a COO matrix/array")
-        elif self.ndim == 1 and other.ndim == 1:
-            # Handle inner product of vectors (1-D arrays)
-            if self.shape[0] != other.shape[0]:
-                raise ValueError(f"shapes {self.shape} and {other.shape}"
-                                 " are not aligned for inner product")
-            return self @ other
-        elif self.ndim == 2 and other.ndim == 2:
-            # Handle matrix multiplication (2-D arrays)
-            if self.shape[1] != other.shape[0]:
-                raise ValueError(f"shapes {self.shape} and {other.shape}"
-                                 " are not aligned for matmul")
-            return self @ other
-        else:
-            return self._sparse_dot(other)
-
+        return self._sparse_dot(other.tocoo())
 
     def _sparse_dot(self, other):
-        self_is_1d = self.ndim == 1
-        other_is_1d = other.ndim == 1
-
-        # reshape to 2-D if self or other is 1-D
-        if self_is_1d:
-            self = self.reshape(self._shape_as_2d)  # prepend 1 to shape
-        if other_is_1d:
-            other = other.reshape((other.shape[0], 1))  # append 1 to shape
-
-        if self.shape[-1] != other.shape[-2]:
-                raise ValueError(f"shapes {self.shape} and {other.shape}"
-                                 " are not aligned for n-D dot")
-
-        # Prepare the tensors for dot operation
+        # already checked: at least one is >2d, neither scalar, both are coo
         # Ravel non-reduced axes coordinates
-        self_raveled_coords = _ravel_non_reduced_axes(self.coords,
-                                                      self.shape, [self.ndim-1])
-        other_raveled_coords = _ravel_non_reduced_axes(other.coords,
-                                                       other.shape, [other.ndim-2])
+        self_2d, s_new_shape = _convert_to_2d(self, [self.ndim - 1])
+        other_2d, o_new_shape = _convert_to_2d(other, [max(0, other.ndim - 2)])
 
-        # Get the shape of the non-reduced axes
-        self_nonreduced_shape = self.shape[:-1]
-        other_nonreduced_shape = other.shape[:-2] + other.shape[-1:]
+        prod = self_2d @ other_2d.T  # routes via 2-D CSR
+        prod = prod.tocoo()
 
-        # Create 2D coords arrays
-        ravel_coords_shape_self = (math.prod(self_nonreduced_shape), self.shape[-1])
-        ravel_coords_shape_other = (other.shape[-2], math.prod(other_nonreduced_shape))
-
-        self_2d_coords = (self_raveled_coords, self.coords[-1])
-        other_2d_coords = (other.coords[-2], other_raveled_coords)
-
-        self_2d = coo_array((self.data, self_2d_coords), ravel_coords_shape_self)
-        other_2d = coo_array((other.data, other_2d_coords), ravel_coords_shape_other)
-
-        prod = (self_2d @ other_2d).tocoo() # routes via 2-D CSR
-
-        # Combine the shapes of the non-reduced axes
-        combined_shape = self_nonreduced_shape + other_nonreduced_shape
+        # Combine the shapes of the non-contracted axes
+        combined_shape = s_new_shape + o_new_shape
 
         # Unravel the 2D coordinates to get multi-dimensional coordinates
-        shapes = (self_nonreduced_shape, other_nonreduced_shape)
-        prod_coords = []
-        for c, s in zip(prod.coords, shapes):
-            prod_coords.extend(np.unravel_index(c, s))
+        coords = []
+        new_shapes = (s_new_shape, o_new_shape) if s_new_shape else (o_new_shape,)
+        for c, s in zip(prod.coords, new_shapes):
+            coords.extend(np.unravel_index(c, s))
 
-        prod_arr = coo_array((prod.data, prod_coords), combined_shape)
-
-        # reshape back if a or b were originally 1-D
-        # TODO: Move this logic before computation of prod_coords for efficiency
-        if self_is_1d:
-            prod_arr = prod_arr.reshape(combined_shape[1:])
-        if other_is_1d:
-            prod_arr = prod_arr.reshape(combined_shape[:-1])
-
-        return prod_arr
+        # Construct the resulting COO array with coords and shape
+        return coo_array((prod.data, coords), shape=combined_shape)
 
     def _dense_dot(self, other):
-        self_is_1d = self.ndim == 1
-        other_is_1d = other.ndim == 1
+        # already checked: self is >0d, other is dense and >0d
+        # Ravel non-reduced axes coordinates
+        s_ndim = self.ndim
+        if s_ndim <= 2:
+            s_new_shape = () if s_ndim == 1 else (self.shape[0],)
+            self_2d = self
+        else:
+            self_2d, s_new_shape = _convert_to_2d(self, [self.ndim - 1])
 
-        # reshape to 2-D if self or other is 1-D
-        if self_is_1d:
-            self = self.reshape(self._shape_as_2d)  # prepend 1 to shape
-        if other_is_1d:
-            other = other.reshape((other.shape[0], 1))  # append 1 to shape
+        o_ndim = other.ndim
+        if o_ndim <= 2:
+            o_new_shape = () if o_ndim == 1 else (other.shape[-1],)
+            other_2d = other
+        else:
+            o_new_shape = other.shape[:-2] + other.shape[-1:]
+            reorder_dims = (o_ndim - 2, *range(o_ndim - 2), o_ndim - 1)
+            o_reorg = np.permute_dims(other, reorder_dims)
+            other_2d = o_reorg.reshape((other.shape[-2], math.prod(o_new_shape)))
 
-        if self.shape[-1] != other.shape[-2]:
-                raise ValueError(f"shapes {self.shape} and {other.shape}"
-                                 " are not aligned for n-D dot")
+        prod = self_2d @ other_2d  # routes via 2-D CSR
 
-        new_shape_self = (
-            self.shape[:-1] + (1,) * (len(other.shape) - 1) + self.shape[-1:]
-        )
-        new_shape_other = (1,) * (len(self.shape) - 1) + other.shape
+        # Combine the shapes of the non-contracted axes
+        combined_shape = s_new_shape + o_new_shape
+        return prod.reshape(combined_shape)
 
-        result_shape = self.shape[:-1] + other.shape[:-2] + other.shape[-1:]
-        result = self.reshape(new_shape_self) @ other.reshape(new_shape_other)
-        prod_arr = result.reshape(result_shape)
-
-        # reshape back if a or b were originally 1-D
-        if self_is_1d:
-            prod_arr = prod_arr.reshape(result_shape[1:])
-        if other_is_1d:
-            prod_arr = prod_arr.reshape(result_shape[:-1])
-
-        return prod_arr
 
     def tensordot(self, other, axes=2):
         """Return the tensordot product with another array along the given axes.
@@ -1097,85 +1041,52 @@ class _coo_base(_data_matrix, _minmax_mixin):
         else:
             return self._sparse_tensordot(other, axes_self, axes_other)
 
-
-    def _sparse_tensordot(self, other, axes_self, axes_other):
-        ndim_self = len(self.shape)
-        ndim_other = len(other.shape)
-
+    def _sparse_tensordot(self, other, s_axes, o_axes):
         # Prepare the tensors for tensordot operation
         # Ravel non-reduced axes coordinates
-        self_non_red_coords = _ravel_non_reduced_axes(self.coords, self.shape,
-                                                      axes_self)
-        self_reduced_coords = np.ravel_multi_index(
-            [self.coords[ax] for ax in axes_self], [self.shape[ax] for ax in axes_self])
-        other_non_red_coords = _ravel_non_reduced_axes(other.coords, other.shape,
-                                                       axes_other)
-        other_reduced_coords = np.ravel_multi_index(
-            [other.coords[a] for a in axes_other], [other.shape[a] for a in axes_other]
-        )
-        # Get the shape of the non-reduced axes
-        self_nonreduced_shape = tuple(self.shape[ax] for ax in range(ndim_self)
-                              if ax not in axes_self)
-        other_nonreduced_shape = tuple(other.shape[ax] for ax in range(ndim_other)
-                               if ax not in axes_other)
-
-        # Create 2D coords arrays
-        ravel_coords_shape_self = (math.prod(self_nonreduced_shape),
-                                math.prod([self.shape[ax] for ax in axes_self]))
-        ravel_coords_shape_other = (math.prod([other.shape[ax] for ax in axes_other]),
-                                    math.prod(other_nonreduced_shape))
-
-        self_2d_coords = (self_non_red_coords, self_reduced_coords)
-        other_2d_coords = (other_reduced_coords, other_non_red_coords)
-
-        self_2d = coo_array((self.data, self_2d_coords), ravel_coords_shape_self)
-        other_2d = coo_array((other.data, other_2d_coords), ravel_coords_shape_other)
+        self_2d, s_new_shape = _convert_to_2d(self, s_axes)
+        other_2d, o_new_shape = _convert_to_2d(other, o_axes)
 
         # Perform matrix multiplication (routed via 2-D CSR)
-        prod = (self_2d @ other_2d).tocoo()
+        prod = self_2d @ other_2d.T
+        # handle case of scalar result (axis includes all axes for both)
+        if not issparse(prod):
+            return prod
+        prod = prod.tocoo()
 
         # Combine the shapes of the non-contracted axes
-        combined_shape = self_nonreduced_shape + other_nonreduced_shape
+        combined_shape = s_new_shape + o_new_shape
 
         # Unravel the 2D coordinates to get multi-dimensional coordinates
         coords = []
-        for c, s in zip(prod.coords, (self_nonreduced_shape, other_nonreduced_shape)):
+        new_shapes = (s_new_shape, o_new_shape) if s_new_shape else (o_new_shape,)
+        for c, s in zip(prod.coords, new_shapes):
             if s:
                 coords.extend(np.unravel_index(c, s))
 
-        if coords == []:  # if result is scalar
-            return sum(prod.data)
-
-        # Construct the resulting COO array with combined coordinates and shape
+        # Construct the resulting COO array with coords and shape
         return coo_array((prod.data, coords), shape=combined_shape)
 
+    def _dense_tensordot(self, other, s_axes, o_axes):
+        s_ndim = len(self.shape)
+        o_ndim = len(other.shape)
 
-    def _dense_tensordot(self, other, axes_self, axes_other):
-        ndim_self = len(self.shape)
-        ndim_other = len(other.shape)
+        s_non_axes = [i for i in range(s_ndim) if i not in s_axes]
+        s_axes_shape = [self.shape[i] for i in s_axes]
+        s_non_axes_shape = [self.shape[i] for i in s_non_axes]
 
-        non_reduced_axes_self = [ax for ax in range(ndim_self) if ax not in axes_self]
-        reduced_shape_self = [self.shape[s] for s in axes_self]
-        non_reduced_shape_self = [self.shape[s] for s in non_reduced_axes_self]
+        o_non_axes = [i for i in range(o_ndim) if i not in o_axes]
+        o_axes_shape = [other.shape[i] for i in o_axes]
+        o_non_axes_shape = [other.shape[i] for i in o_non_axes]
 
-        non_reduced_axes_other = [ax for ax in range(ndim_other)
-                                  if ax not in axes_other]
-        reduced_shape_other = [other.shape[s] for s in axes_other]
-        non_reduced_shape_other = [other.shape[s] for s in non_reduced_axes_other]
+        left = self.transpose(s_non_axes + s_axes)
+        right = np.transpose(other, o_non_axes[:-1] + o_axes + o_non_axes[-1:])
 
-        permute_self = non_reduced_axes_self + axes_self
-        permute_other = (
-            non_reduced_axes_other[:-1] + axes_other + non_reduced_axes_other[-1:]
-        )
-        self = self.transpose(permute_self)
-        other = np.transpose(other, permute_other)
+        reshape_left = (*s_non_axes_shape, math.prod(s_axes_shape))
+        reshape_right = (*o_non_axes_shape[:-1], math.prod(o_axes_shape),
+                         *o_non_axes_shape[-1:])
 
-        reshape_self = (*non_reduced_shape_self, math.prod(reduced_shape_self))
-        reshape_other = (*non_reduced_shape_other[:-1], math.prod(reduced_shape_other),
-                        *non_reduced_shape_other[-1:])
-
-        return self.reshape(reshape_self).dot(other.reshape(reshape_other))
-
+        return left.reshape(reshape_left).dot(right.reshape(reshape_right))
 
     def _matmul_sparse(self, other):
         """
@@ -1218,7 +1129,6 @@ class _coo_base(_data_matrix, _minmax_mixin):
             prod_block_diag,
             shape=(*broadcast_shape, self.shape[-2], other.shape[-1]),
         )
-
 
     def _broadcast_to(self, new_shape, copy=False):
         if self.shape == new_shape:
@@ -1274,6 +1184,26 @@ class _coo_base(_data_matrix, _minmax_mixin):
                 new_coords = (new_dim,) + new_coords
 
         return coo_array((new_data, new_coords), new_shape)
+
+    def _sum_nd(self, axis, res_dtype, out):
+        # axis and out are preprocessed. out.shape is new_shape
+        A2d, new_shape = _convert_to_2d(self, axis)
+        ones = np.ones((A2d.shape[1], 1), dtype=res_dtype)
+        # sets dtype while loading into out
+        out[...] = (A2d @ ones).reshape(new_shape)
+        return out
+
+    def _min_or_max_axis_nd(self, axis, min_or_max, explicit):
+        A2d, new_shape = _convert_to_2d(self, axis)
+        res = A2d._min_or_max_axis(1, min_or_max, explicit)
+        unraveled_coords = np.unravel_index(res.coords[0], new_shape)
+
+        return coo_array((res.data, unraveled_coords), new_shape)
+
+    def _argminmax_axis_nd(self, axis, argminmax, compare, explicit):
+        A2d, new_shape = _convert_to_2d(self, axis)
+        res_flat = A2d._argminmax_axis(1, argminmax, compare, explicit)
+        return res_flat.reshape(new_shape)
 
 
 def _block_diag(self):
@@ -1349,22 +1279,26 @@ def _process_axes(ndim_a, ndim_b, axes):
     return axes_a, axes_b
 
 
-def _ravel_non_reduced_axes(coords, shape, axes):
-    ndim = len(shape)
-    non_reduced_axes = [ax for ax in range(ndim) if ax not in axes]
+def _convert_to_2d(coo, axis):
+    axis_coords = tuple(coo.coords[i] for i in axis)
+    axis_shape = tuple(coo.shape[i] for i in axis)
+    axis_ravel = _ravel_coords(axis_coords, axis_shape)
 
-    if not non_reduced_axes:
-        # Return an array with one row
-        return np.zeros_like(coords[0])
+    ndim = len(coo.coords)
+    non_axis = tuple(i for i in range(ndim) if i not in axis)
+    if non_axis:
+        non_axis_coords = tuple(coo.coords[i] for i in non_axis)
+        non_axis_shape = tuple(coo.shape[i] for i in non_axis)
+        non_axis_ravel = _ravel_coords(non_axis_coords, non_axis_shape)
+        coords_2d = (non_axis_ravel, axis_ravel)
+        shape_2d = (math.prod(non_axis_shape), math.prod(axis_shape))
+    else:  # all axes included in axis so result will have 1 element
+        coords_2d = (axis_ravel,)
+        shape_2d = (math.prod(axis_shape),)
+        non_axis_shape = ()
 
-    # Extract the shape of the non-reduced axes
-    non_reduced_shape = [shape[ax] for ax in non_reduced_axes]
-
-    # Extract the coordinates of the non-reduced axes
-    non_reduced_coords = tuple(coords[idx] for idx in non_reduced_axes)
-
-    # Ravel the coordinates into 1D
-    return np.ravel_multi_index(non_reduced_coords, non_reduced_shape)
+    new_coo = coo_array((coo.data, coords_2d), shape=shape_2d)
+    return new_coo, non_axis_shape
 
 
 def _ravel_coords(coords, shape, order='C'):

--- a/scipy/sparse/_data.py
+++ b/scipy/sparse/_data.py
@@ -124,7 +124,7 @@ class _data_matrix(_spbase):
 
         data = self._deduped_data()
         if dtype is not None:
-            data = data.astype(dtype)
+            data = data.astype(dtype, copy=False)
         return self._with_data(data ** n)
 
     ###########################
@@ -172,9 +172,8 @@ class _minmax_mixin:
     """
 
     def _min_or_max_axis(self, axis, min_or_max, explicit):
+        # already checked that self.shape[axis] is not zero
         N = self.shape[axis]
-        if N == 0:
-            raise ValueError("zero-size array to reduction operation")
         M = self.shape[1 - axis]
         idx_dtype = self._get_index_dtype(maxval=M)
 
@@ -208,7 +207,7 @@ class _minmax_mixin:
 
     def _min_or_max(self, axis, out, min_or_max, explicit):
         if out is not None:
-            raise ValueError("Sparse arrays do not support an 'out' parameter.")
+            raise ValueError("Sparse min/max does not support an 'out' parameter.")
 
         axis = validateaxis(axis, ndim=self.ndim)
 
@@ -224,12 +223,15 @@ class _minmax_mixin:
                 m = min_or_max(zero, m)
             return m
 
-        return self._min_or_max_axis(axis, min_or_max, explicit)
+        if any(self.shape[d] == 0 for d in axis):
+            raise ValueError("zero-size array to reduction operation")
 
-    def _arg_min_or_max_axis(self, axis, argmin_or_argmax, compare, explicit):
-        if self.shape[axis] == 0:
-            raise ValueError("Cannot apply the operation along a zero-sized dimension.")
+        if self.ndim == 2:
+            # note: 2D ensures that len(axis)==1 so we pass in the int axis[0]
+            return self._min_or_max_axis(axis[0], min_or_max, explicit)
+        return self._min_or_max_axis_nd(axis, min_or_max, explicit)
 
+    def _argminmax_axis(self, axis, argminmax, compare, explicit):
         zero = self.dtype.type(0)
 
         mat = self.tocsc() if axis == 0 else self.tocsr()
@@ -243,7 +245,7 @@ class _minmax_mixin:
             p, q = mat.indptr[i:i + 2]
             data = mat.data[p:q]
             indices = mat.indices[p:q]
-            extreme_index = argmin_or_argmax(data)
+            extreme_index = argminmax(data)
             extreme_value = data[extreme_index]
             if explicit:
                 if q - p > 0:
@@ -266,21 +268,31 @@ class _minmax_mixin:
 
         return self._ascontainer(ret)
 
-    def _arg_min_or_max(self, axis, out, argmin_or_argmax, compare, explicit):
+    def _argminmax(self, axis, out, argminmax, compare, explicit):
         if out is not None:
-            raise ValueError("Sparse types do not support an 'out' parameter.")
+            minmax = "argmin" if argminmax == np.argmin else "argmax"
+            raise ValueError(f"Sparse {minmax} does not support an 'out' parameter.")
 
         axis = validateaxis(axis, ndim=self.ndim)
 
         if axis is not None:
-            return self._arg_min_or_max_axis(axis, argmin_or_argmax, compare, explicit)
+            if any(self.shape[i] == 0 for i in axis):
+                minmax = "argmin" if argminmax == np.argmin else "argmax"
+                raise ValueError(f"Cannot apply {minmax} along a zero-sized dimension.")
+
+            if self.ndim == 2:
+                # note: 2D ensures that len(axis)==1 so we pass in the int axis[0]
+                return self._argminmax_axis(axis[0], argminmax, compare, explicit)
+            return self._argminmax_axis_nd(axis, argminmax, compare, explicit)
 
         if 0 in self.shape:
-            raise ValueError("Cannot apply the operation to an empty matrix.")
+            minmax = "argmin" if argminmax == np.argmin else "argmax"
+            raise ValueError(f"Cannot apply {minmax} to an empty matrix.")
 
         if self.nnz == 0:
             if explicit:
-                raise ValueError("Cannot apply the operation to zero matrix "
+                minmax = "argmin" if argminmax == np.argmin else "argmax"
+                raise ValueError(f"Cannot apply {minmax} to zero matrix "
                                  "when explicit=True.")
             return 0
 
@@ -288,28 +300,34 @@ class _minmax_mixin:
         mat = self.tocoo()
         # Convert to canonical form: no duplicates, sorted indices.
         mat.sum_duplicates()
-        extreme_index = argmin_or_argmax(mat.data)
+        extreme_index = argminmax(mat.data)
         if explicit:
             return extreme_index
         extreme_value = mat.data[extreme_index]
-        num_col = mat.shape[-1]
 
+        if mat.ndim > 2:
+            mat = mat.reshape(-1)
         # If the min value is less than zero, or max is greater than zero,
         # then we do not need to worry about implicit zeros.
-        if compare(extreme_value, zero):
+        # And we use a "cheap test" for the rare case of no implicit zeros.
+        maxnnz = math.prod(self.shape)
+        if compare(extreme_value, zero) or mat.nnz == maxnnz:
             # cast to Python int to avoid overflow and RuntimeError
-            return int(mat.row[extreme_index]) * num_col + int(mat.col[extreme_index])
-
-        # Cheap test for the rare case where we have no implicit zeros.
-        size = math.prod(self.shape)
-        if size == mat.nnz:
+            if mat.ndim == 1:  # includes nD case that was reshaped above
+                return int(mat.col[extreme_index])
+            # ndim == 2
+            num_col = mat.shape[-1]
             return int(mat.row[extreme_index]) * num_col + int(mat.col[extreme_index])
 
         # At this stage, any implicit zero could be the min or max value.
         # After sum_duplicates(), the `row` and `col` arrays are guaranteed to
         # be sorted in C-order, which means the linearized indices are sorted.
-        linear_indices = mat.row * num_col + mat.col
-        first_implicit_zero_index = _find_missing_index(linear_indices, size)
+        if mat.ndim == 1:  # includes nD case that was reshaped above
+            linear_indices = mat.coords[-1]
+        else:  # ndim == 2
+            num_col = mat.shape[-1]
+            linear_indices = mat.row * num_col + mat.col
+        first_implicit_zero_index = _find_missing_index(linear_indices, maxnnz)
         if extreme_value == zero:
             return min(first_implicit_zero_index, extreme_index)
         return first_implicit_zero_index
@@ -506,7 +524,7 @@ class _minmax_mixin:
 
         explicit : {False, True} optional (default: False)
             When set to True, only explicitly stored elements will be considered.
-            If axis is not None and a row/column has no stored elements, argmax
+            If axis is not None and an axis has no stored elements, argmax
             is undefined, so the index ``0`` is returned for that row/column.
 
             .. versionadded:: 1.15.0
@@ -516,7 +534,7 @@ class _minmax_mixin:
         ind : numpy.matrix or int
             Indices of maximum elements. If matrix, its size along `axis` is 1.
         """
-        return self._arg_min_or_max(axis, out, np.argmax, np.greater, explicit)
+        return self._argminmax(axis, out, np.argmax, np.greater, explicit)
 
     def argmin(self, axis=None, out=None, *, explicit=False):
         """Return indices of minimum elements along an axis.
@@ -538,7 +556,7 @@ class _minmax_mixin:
 
         explicit : {False, True} optional (default: False)
             When set to True, only explicitly stored elements will be considered.
-            If axis is not None and a row/column has no stored elements, argmin
+            If axis is not None and an axis has no stored elements, argmin
             is undefined, so the index ``0`` is returned for that row/column.
 
             .. versionadded:: 1.15.0
@@ -548,4 +566,4 @@ class _minmax_mixin:
          ind : numpy.matrix or int
             Indices of minimum elements. If matrix, its size along `axis` is 1.
         """
-        return self._arg_min_or_max(axis, out, np.argmin, np.less, explicit)
+        return self._argminmax(axis, out, np.argmin, np.less, explicit)

--- a/scipy/sparse/_dia.py
+++ b/scipy/sparse/_dia.py
@@ -148,7 +148,7 @@ class _dia_base(_data_matrix):
         num_rows, num_cols = self.shape
         ret = None
 
-        if axis == 0:
+        if axis == (0,):
             mask = self._data_mask()
             x = (self.data * mask).sum(axis=0)
             if x.shape[0] == num_cols:
@@ -158,7 +158,7 @@ class _dia_base(_data_matrix):
                 res[:x.shape[0]] = x
             ret = self._ascontainer(res, dtype=res_dtype)
 
-        else:
+        else:  # axis is None or (1,)
             row_sums = np.zeros((num_rows, 1), dtype=res_dtype)
             one = np.ones(num_cols, dtype=res_dtype)
             dia_matvec(num_rows, num_cols, len(self.offsets),

--- a/scipy/sparse/_sputils.py
+++ b/scipy/sparse/_sputils.py
@@ -437,8 +437,6 @@ def validateaxis(axis, *, ndim=2) -> int | tuple[int, ...]:
         raise ValueError("axis tuple has too many elements")
     elif len_axis == ndim:
         return None
-    elif len_axis == 1:
-        return canon_axis[0]
     else:
         return tuple(canon_axis)
 

--- a/scipy/sparse/tests/test_arithmetic1d.py
+++ b/scipy/sparse/tests/test_arithmetic1d.py
@@ -320,7 +320,7 @@ class TestArithmetic1D:
         assert_equal(asp.dot(asp), np.dot(a, a))
 
         # bad matrix products
-        with pytest.raises(ValueError, match='dimension mismatch'):
+        with pytest.raises(ValueError, match='dimension mismatch|shapes.*not aligned'):
             asp.dot(f)
 
         # elemente-wise multiplication

--- a/scipy/sparse/tests/test_base.py
+++ b/scipy/sparse/tests/test_base.py
@@ -1053,7 +1053,9 @@ class _TestCommon:
             datsp.sum(axis=(0, 3))
         with assert_raises(TypeError, match="axis must be an integer"):
             datsp.sum(axis=1.5)
-        assert_raises(ValueError, datsp.sum, axis=1, out=out)
+        # error msg varies by sparray (1d result) or spmatrix (2d result)
+        with assert_raises(ValueError, match="do.*n.t match.*shape|wrong.*dimensions"):
+            datsp.mean(axis=1, out=out)
 
     def test_sum_dtype(self):
         dat = array([[0, 1, 2],
@@ -1151,7 +1153,7 @@ class _TestCommon:
         for dtype in self.checked_dtypes:
             check(dtype)
 
-    def test_mean_invalid_params(self):
+    def test_mean_invalid_param(self):
         out = self.asdense(np.zeros((1, 3)))
         dat = array([[0, 1, 2],
                      [3, -4, 5],
@@ -1164,7 +1166,8 @@ class _TestCommon:
             datsp.mean(axis=(0, 3))
         with assert_raises(TypeError, match="axis must be an integer"):
             datsp.mean(axis=1.5)
-        with assert_raises(ValueError, match="doesn't match.*shape|wrong.*dimensions"):
+        # error msg varies by sparray (1d result) or spmatrix (2d result)
+        with assert_raises(ValueError, match="do.*n.t match.*shape|wrong.*dimensions"):
             datsp.mean(axis=1, out=out)
 
     def test_mean_dtype(self):
@@ -3722,9 +3725,11 @@ class _TestMinMax:
         X = self.spcreator(np.arange(1, 10).reshape(3, 3))
         assert_equal(X.min(), 1)
         assert_equal(X.min().dtype, X.dtype)
+        assert_equal(X.min(explicit=True), 1)
 
         X = -X
         assert_equal(X.max(), -1)
+        assert_equal(X.max(explicit=True), -1)
 
         # and a fully sparse matrix
         Z = self.spcreator(np.zeros((1, 1)))
@@ -3886,10 +3891,12 @@ class _TestMinMax:
         datsp = self.spcreator(dat)
 
         for fname in ('min', 'max'):
+            datfunc = getattr(dat, fname)
             func = getattr(datsp, fname)
             assert_raises(ValueError, func, axis=3)
             assert_raises(TypeError, func, axis=1.5)
             assert_raises(ValueError, func, axis=1, out=1)
+            assert_equal(func(axis=(0, 1)), datfunc(axis=(0, 1)))
 
     def test_numpy_minmax(self):
         # See gh-5987
@@ -3938,6 +3945,13 @@ class _TestMinMax:
 
             assert_equal(mat.argmax(axis=1), np.argmax(D, axis=1))
             assert_equal(mat.argmin(axis=1), np.argmin(D, axis=1))
+
+        # full matrix with explicit=True
+        mat = self.spcreator(self.asdense(D5))
+        assert_equal(mat.argmax(explicit=True), 5)
+        assert_equal((-mat).argmax(explicit=True), 2)
+        assert_equal(mat.argmin(explicit=True), 2)
+        assert_equal((-mat).argmin(explicit=True), 5)
 
         # zero-size matrices
         D6 = self.spcreator(np.empty((0, 5)))

--- a/scipy/sparse/tests/test_common1d.py
+++ b/scipy/sparse/tests/test_common1d.py
@@ -180,7 +180,7 @@ class TestCommon1D:
             datsp.mean(axis=(0, 3))
         with pytest.raises(TypeError, match='axis must be an integer'):
             datsp.mean(axis=1.5)
-        with pytest.raises(ValueError, match='output parameter.*wrong.*dimension'):
+        with pytest.raises(ValueError, match='out.*not match shape'):
             datsp.mean(axis=1, out=out)
 
     def test_sum_dtype(self, spcreator):

--- a/scipy/sparse/tests/test_coo.py
+++ b/scipy/sparse/tests/test_coo.py
@@ -1,5 +1,5 @@
 import numpy as np
-from numpy.testing import assert_equal
+from numpy.testing import assert_equal, assert_allclose
 import pytest
 from scipy.linalg import block_diag
 from scipy.sparse import coo_array, random_array
@@ -563,8 +563,7 @@ def test_nd_add_sparse_with_inconsistent_shapes(a_shape, b_shape):
 
     arr_a = random_array((a_shape), density=0.6, rng=rng, dtype=int)
     arr_b = random_array((b_shape), density=0.6, rng=rng, dtype=int)
-    with pytest.raises(ValueError,
-                       match="(Incompatible|inconsistent) shapes|cannot be broadcast"):
+    with pytest.raises(ValueError, match="inconsistent shapes"):
         arr_a + arr_b
 
 
@@ -849,3 +848,134 @@ def test_extract_block_diag(shape):
     res = _extract_block_diag(_block_diag(sp_x), shape)
 
     assert_equal(res.toarray(), sp_x.toarray())
+
+
+add_sub_shapes = [
+    ((3,4), (3,4)), ((3,4,6), (3,4,6)), ((3,7,5), (3,7,5))
+]
+@pytest.mark.parametrize(('a_shape', 'b_shape'), add_sub_shapes)
+def test_add_no_broadcasting(a_shape, b_shape):
+    rng = np.random.default_rng(23409823)
+    a = random_array(a_shape, density=0.6, random_state=rng, dtype=int)
+    b = random_array(b_shape, density=0.6, random_state=rng, dtype=int)
+
+    res = a + b
+    exp = np.add(a.toarray(), b.toarray())
+    assert_equal(res.toarray(), exp)
+    res = a + b.toarray()
+    assert_equal(res, exp)
+
+@pytest.mark.parametrize(('a_shape', 'b_shape'), add_sub_shapes)
+def test_sub_no_broadcasting(a_shape, b_shape):
+    rng = np.random.default_rng(23409823)
+    a = random_array(a_shape, density=0.6, random_state=rng, dtype=int)
+    b = random_array(b_shape, density=0.6, random_state=rng, dtype=int)
+
+    res = a - b
+    exp = np.subtract(a.toarray(), b.toarray())
+    assert_equal(res.toarray(), exp)
+
+    res = a - b.toarray()
+    assert_equal(res, exp)
+
+argmax_argmin_shapes_axis = [
+    ((3,), None), ((3,), 0),
+    ((4,6), 1), ((7,3), 0), ((3,5), None),
+    ((2,8,7), 2), ((2,8,7), 0),
+    ((2,0), 0), ((3,0,0,2), 0),
+    ((3,2,4,7), None), ((3,2,4,7), 1), ((3,2,4,7), 0), ((3,2,4,7), 2),
+    ((3,2,4,7), -2), ((4,5,7,8,2), 4), ((4,5,7,8,2), -3),
+]
+@pytest.mark.parametrize(('shape', 'axis'), argmax_argmin_shapes_axis)
+def test_argmax_argmin(shape, axis):
+    rng = np.random.default_rng(23409823)
+    a = random_array(shape, density=0.6, random_state=rng, dtype=int)
+
+    res = a.argmax(axis=axis)
+    exp = np.argmax(a.toarray(), axis=axis)
+    assert_equal(res, exp)
+
+    res = a.argmin(axis=axis)
+    exp = np.argmin(a.toarray(), axis=axis)
+    assert_equal(res, exp)
+
+
+max_min_shapes_axis = [
+    ((3,), None), ((3,), 0),
+    ((4,6), 1), ((7,3), 0), ((3,5), None),
+    ((2,8,7), 2), ((2,8,7), 0),
+    ((3,2,4,7), None), ((3,2,4,7), 1), ((3,2,4,7), 0), ((3,2,4,7), 2),
+    ((4,5,7,8,2), 4), ((4,5,8,1), 3), ((4,6), (0,)), ((4,6), (0,1)),
+    ((3,0,2), 2), ((3,0,2), (0,2)), ((3,0), 0),
+    ((3,7,8,5), (0,1)), ((3,7,8,5), (2,1)), ((3,7,8,5), (2,0)),
+    ((3,7,8,5), (0,-2)), ((3,7,8,5), (-1,2)), ((3,7,8,5), (3)),
+    ((3,7,8,5), (0,1,2)), ((3,7,8,5), (0,1,2,3)),
+]
+@pytest.mark.parametrize(('shape', 'axis'), max_min_shapes_axis)
+def test_min_max(shape, axis):
+    rng = np.random.default_rng(23409823)
+    a = random_array(shape, density=0.6, random_state=rng, dtype=int)
+
+    res_min = a.min(axis=axis)
+    exp_min = np.min(a.toarray(), axis=axis)
+    res_max = a.max(axis=axis)
+    exp_max = np.max(a.toarray(), axis=axis)
+    res_nanmin = a.nanmin(axis=axis)
+    exp_nanmin = np.nanmin(a.toarray(), axis=axis)
+    res_nanmax = a.nanmax(axis=axis)
+    exp_nanmax = np.nanmax(a.toarray(), axis=axis)
+
+    for res, exp in [(res_min, exp_min), (res_max, exp_max),
+                     (res_nanmin, exp_nanmin), (res_nanmax, exp_nanmax)]:
+        if np.issubdtype(type(res), np.number):
+            assert_equal(res, exp)
+        else:
+            assert_equal(res.toarray(), exp)
+
+
+def test_min_max_full():
+    for a in (coo_array([[[1, 2, 3, 4]]]), coo_array([[1, 2, 3, 4]])):
+        assert a.min() == 1
+        assert (-a).max() == -1
+
+
+sum_mean_params = [
+    ((3,), None, None), ((3,), 0, None),
+    ((4,6), 1, None), ((7,3), 0, None), ((3,5), None, None),
+    ((2,8,7), 2, None), ((2,8,7), 0, np.zeros((8,7))),
+    ((3,2,4,7), None, None), ((3,2,4,7), 1, np.zeros((3,4,7))),
+    ((3,2,4,7), 0, None), ((4,5,7,8,2), 4, None),
+    ((4,5,8,1), 3, None), ((4,6), (0,), None), ((4,6), (0,1), None),
+    ((3,0,2), 2, None), ((3,0,2), (0,2), None), ((3,0), 0, None),
+    ((3,7,8,5), (0,1), np.zeros((8,5))), ((3,7,8,5), (2,1), None),
+    ((3,7,8,5), (0,-2), None), ((3,7,8,5), (-1,2), np.zeros((3,7))),
+    ((3,7,8,5), (3), None), ((3,7,8,5), (0,1,2), np.zeros((5,))),
+    ((3,7,8,5), (0,1,2,3), None),
+]
+@pytest.mark.parametrize(('shape', 'axis', 'out'), sum_mean_params)
+def test_sum(shape, axis, out):
+    rng = np.random.default_rng(23409823)
+    a = random_array(shape, density=0.6, random_state=rng, dtype=int)
+
+    res = a.sum(axis=axis, out=out)
+    exp = np.sum(a.toarray(), axis=axis, out=out)
+    assert_equal(res, exp)
+
+
+@pytest.mark.parametrize(('shape', 'axis', 'out'), sum_mean_params)
+def test_mean(shape, axis, out):
+    rng = np.random.default_rng(23409823)
+    a = random_array(shape, density=0.6, random_state=rng, dtype=int)
+
+    res = a.mean(axis=axis, out=out)
+    exp = np.mean(a.toarray(), axis=axis, out=out)
+    assert_allclose(res, exp)
+
+
+def test_pow_abs_round():
+    rng = np.random.default_rng(23409823)
+    a = random_array((3,6,5,2,4), density=0.6, random_state=rng, dtype=int)
+    assert_allclose((a**3).toarray(), np.power(a.toarray(), 3))
+    assert_allclose((a**7).toarray(), np.power(a.toarray(), 7))
+    assert_allclose(round(a).toarray(), np.round(a.toarray()))
+    assert_allclose(abs(a).toarray(), np.abs(a.toarray()))

--- a/scipy/sparse/tests/test_sputils.py
+++ b/scipy/sparse/tests/test_sputils.py
@@ -131,19 +131,18 @@ class TestSparseUtils:
         for ax in [5, -5, (0, 5), (-5, 0)]:
             with assert_raises(ValueError, match="out of range"):
                 sputils.validateaxis(ax, ndim=2)
-        for axis in (0, 1, None):
+        for axis in ((0,), (1,), None):
             assert sputils.validateaxis(axis, ndim=2) == axis
-        convert_axis_2d = {-2: 0, -1: 1, (0, 1): None, (0, -1): None}
-        for axis, canonical_axis in convert_axis_2d.items():
+        axis_2d = {-2: (0,), -1: (1,), 0: (0,), 1: (1,), (0, 1): None, (0, -1): None}
+        for axis, canonical_axis in axis_2d.items():
             assert sputils.validateaxis(axis, ndim=2) == canonical_axis
 
         # ndim 4
-        for axis in (2, 3, (2, 3), (2, 1), (0, 3)):
+        for axis in ((2,), (3,), (2, 3), (2, 1), (0, 3)):
             assert sputils.validateaxis(axis, ndim=4) == axis
-        convert_axis_4d = {-4: 0, -3: 1, (3, -4): (3, 0)}
-        for axis, canonical_axis in convert_axis_4d.items():
+        axis_4d = {-4: (0,), -3: (1,), 2: (2,), 3: (3,), (3, -4): (3, 0)}
+        for axis, canonical_axis in axis_4d.items():
             sputils.validateaxis(axis, ndim=4) == canonical_axis
-
 
     @pytest.mark.parametrize("container", [csr_array, bsr_array])
     def test_safely_cast_index_compressed(self, container):

